src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs

```diff
@@ -566,6 +574,203 @@ internal void AssertJsonCondition(params JsonNodeType[] allowedNodeTypes)
+
+            while (propertyParsingResult == PropertyParsingResult.CustomInstanceAnnotation && this.ShouldSkipCustomInstanceAnnotation(propertyName))
+            {
+                // Read over the property name
+                await this.JsonReader.ReadAsync()
+                    .ConfigureAwait(false);
+
+                // Skip over the instance annotation value
+                await this.JsonReader.SkipValueAsync()
+                    .ConfigureAwait(false);
+                parsePropertyResult = await this.ParsePropertyAsync(
+                    propertyAndAnnotationCollector,
+                    readPropertyAnnotationValueDelegate).ConfigureAwait(false);
+                propertyParsingResult = parsePropertyResult.Item1;
+                propertyName = parsePropertyResult.Item2;
+            }
+
+            await handlePropertyDelegate(propertyParsingResult, propertyName)
+                .ConfigureAwait(false);
+            if (propertyParsingResult != PropertyParsingResult.EndOfObject
+                && propertyParsingResult != PropertyParsingResult.CustomInstanceAnnotation)
+            {
+                propertyAndAnnotationCollector.MarkPropertyAsProcessed(propertyName);
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously reads the odata.context annotation.
+        /// </summary>
+        /// <param name="payloadKind">The payload kind for which to read the context URI.</param>
+        /// <param name="propertyAndAnnotationCollector">The duplicate property names checker.</param>
+        /// <param name="failOnMissingContextUriAnnotation">true if the method should fail if the context URI annotation is missing, false if that can be ignored.</param>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the value of the context URI annotation.
+        /// </returns>
+        internal async Task<string> ReadContextUriAnnotationAsync(
+            ODataPayloadKind payloadKind,
+            PropertyAndAnnotationCollector propertyAndAnnotationCollector,
+            bool failOnMissingContextUriAnnotation)
+        {
+            if (this.JsonReader.NodeType != JsonNodeType.Property)
+            {
+                if (!failOnMissingContextUriAnnotation || payloadKind == ODataPayloadKind.Unsupported)
+                {
+                    // Do not fail during payload kind detection
+                    return null;
+                }
+
+                throw new ODataException(Strings.ODataJsonLightDeserializer_ContextLinkNotFoundAsFirstProperty);
+            }
+
+            // Must make sure the input odata.context has a '@' prefix
+            string propertyName = await this.JsonReader.GetPropertyNameAsync()
+                .ConfigureAwait(false);
+            if (string.CompareOrdinal(JsonLightConstants.ODataPropertyAnnotationSeparatorChar + ODataAnnotationNames.ODataContext, propertyName) != 0

```

Comment:
@habbes Because we do this kind of comparison quite a lot, I'm thinking of introducing constants that comprise of the prefixed OData annotation names, i.e. `ODataAnnotationNames.PrefixedODataContext` with the value as `@odata.context`. That will altogether deal with the unnecessary allocation.